Add boxing test

RaduBerinde opened the pull request:
Boxing benchmark which reveals in what cases Go makes an allocation just by
assigning a variable to an interface.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/go-lab/2)

<!-- Reviewable:end -->

RaduBerinde committed:
Add boxing test

Boxing benchmark which reveals in what cases Go makes an allocation just by
assigning a variable to an interface.

nvb commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

_[boxing/main_test.go, line 2 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUFghWlcMISRW71XAr:-KPUFghWlcMISRW71XAs:bi3vcyb) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/main_test.go#L2)):_

> ``` Go
> // Test programs used to determine in what cases assigning an interface to a
> // variable results in Go making an allocation.
> ```

Maybe mention why you're sending each value through a channel.

---

_[boxing/main_test.go, line 4 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUFoJ7Tag1Ho08Tpv5:-KPUFoJ7Tag1Ho08Tpv6:b5ak45s) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/main_test.go#L4)):_

> ``` Go
> // variable results in Go making an allocation.
> //
> // Sample output:
> ```

In the future we could process this output to get something like

```
benchmark      | allocates
Int            | ✓
EmptyStruct    | x
...
```

---

_[boxing/main_test.go, line 28 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUERn-y5qY98OS82sc:-KPUERn-y5qY98OS82sd:bxj5nwi) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/main_test.go#L28)):_

> ``` Go
>   // We use this interface to generate pointer and slice values that can't be
>   // determined at compile time.
>   gen = &VarGenImpl{}
> ```

huh, assigning the implementation to the interface dynamically in the `init` function makes a difference?

---

_[boxing/main_test.go, line 32 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUGSsoxZ8kiXtikpwD:-KPUGSsp09wSmvfAcMmW:b-opbsuj) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/main_test.go#L32)):_

> ``` Go
> 
> func BenchmarkInt(b *testing.B) {
>   ch := make(chan interface{})
> ```

Could most of this shared code be factored out into something like:

```
func sendOnChan(b *testing.B, f func(chan<- interface{})) {
    ch := make(chan interface{})
    go func() {
        for _ = range ch {
        }
    }()
    for i := 0; i < b.N; i++ {
        f(ch)
    }
    close(ch)
}

func BenchmarkInt(b *testing.B) {
    sendOnChan(b, func(c chan<- interface{}) {
        x := int(i * i)
        c <- x
    })
}
```

---

_[boxing/Makefile, line 1 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUFZ-k-u_7jjmIHR_L:-KPUFZ-k-u_7jjmIHR_M:byl52kr) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/Makefile#L1)):_

> ``` Makefile
> all:
> ```

`.PHONY: all`?

---

_[boxing/var_gen.go, line 16 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUFH4No_RkdXMXrMsA:-KPUFH4No_RkdXMXrMsB:bwtsgyn) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/var_gen.go#L16)):_

> ``` Go
> // VarGen implementation
> type VarGenImpl struct {
>   pos        int
> ```

Is there a reason we need to cycle through a small group of values?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/go-lab/2)_

<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

_[boxing/main_test.go, line 28 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUERn-y5qY98OS82sc:-KPULPFFl9d2xOvcnjQE:brst3nw) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/main_test.go#L28)):_
<details><summary>

<i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>



> huh, assigning the implementation to the interface dynamically in the `init` function makes a difference?
> </details>
> Just wrote it like that from the getgo, fixed

---

_[boxing/main_test.go, line 32 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUGSsoxZ8kiXtikpwD:-KPUMf3PhDU4Nuabr2bh:b-dtkdrq) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/main_test.go#L32)):_
<details><summary>

<i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>



> Could most of this shared code be factored out into something like:
> 
> ```
> func sendOnChan(b *testing.B, f func(chan<- interface{})) {
>     ch := make(chan interface{})
>     go func() {
>         for _ = range ch {
>         }
>     }()
>     for i := 0; i < b.N; i++ {
>         f(ch)
>     }
>     close(ch)
> }
> 
> func BenchmarkInt(b *testing.B) {
>     sendOnChan(b, func(c chan<- interface{}) {
>         x := int(i * i)
>         c <- x
>     })
> }
> ```
> 
> </details>
> Done, thanks.

---

_[boxing/Makefile, line 1 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUFZ-k-u_7jjmIHR_L:-KPUNOENCWGu9FvTRFbR:b-896fix) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/Makefile#L1)):_
<details><summary>

<i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>



> `.PHONY: all`?
> </details>
> Done.

---

_[boxing/var_gen.go, line 16 [r1]](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-KPUFH4No_RkdXMXrMsA:-KPUNC8if8GQsirbC-hr:b-kj8xw8) ([raw file](https://github.com/cockroachdb/go-lab/blob/e88ffaae3ef8c8ba7c67a17a74b2aec0f9e34442/boxing/var_gen.go#L16)):_
<details><summary>

<i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>



> Is there a reason we need to cycle through a small group of values?
> </details>
> No, removed.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/go-lab/2)_

<!-- Sent from Reviewable.io -->

nvb commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/go-lab/2#-:-KPUQPybQv1tK29yWkR4:bnfp4nl)_

<!-- Sent from Reviewable.io -->
